File: client-go/kit/protocol.go
package main

// Parse e encode das linhas do protocolo, na fronteira.
// Regra do PROTOCOL.md: tokens extras no final são ignorados (leitor
// tolerante), linha malformada devolve erro e nunca panica.
// Refs: <nome>-<sessão>-<n>, sessão de 4 chars gerada uma vez por processo.
//
// Este arquivo mistura exercício e infra:
//   - E1 (ParseObservation) e E2 (EncodeAction) são stubs seus para implementar.
//   - Todo o resto é INFRA pronta: não edite.

import (
	"fmt"
	"math/rand"
	"strconv"
	"strings"
)

// --- tipos de domínio ------------------------------------------------------

type Cell struct {
	X, Y      int
	Terrain   string
	Owner     int
	Fortified int
}

type ScoreEntry struct {
	ID     int
	Name   string
	Cells  int
	Status string
}

type Change struct {
	X, Y, Owner, Fortified int
}

// Message carrega qualquer mensagem servidor -> cliente; Type discrimina.
type Message struct {
	Type string // "welcome", "obs", "ack", "nack", "err", "score", "diff", "pong"
	Ref  string

	// welcome
	Role       string // "colony" ou "spectator"
	ID         int
	Name       string
	Color      string
	W, H       int
	TickMs     int
	ViewRadius int
	SpawnX     int
	SpawnY     int

	// obs / score / diff
	Tick       int
	Status     string
	ScoresTick int
	Cells      []Cell
	Entries    []ScoreEntry
	Changes    []Change

	// nack / err
	Code   string
	Detail string
}

// Action é a decisão da estratégia, antes do encode.
type Action struct {
	Kind string // "expand", "attack", "fortify", "pass"
	X, Y int
}

// --- encode ----------------------------------------------------------------

type Refs struct {
	name, session string
	n             int
}

// NewRefs cria o gerador de refs de uma sessão.
//
// O nome é o da colônia; a sessão (4 chars aleatórios) é gerada uma vez por
// processo, assim um cliente reiniciado nunca colide na dedup do servidor.
//
// Exemplo:
//
//	refs := NewRefs("aurora")
//	ref := refs.Next() // "aurora-k3f9-1" (sessão aleatória)
func NewRefs(name string) *Refs {
	return &Refs{name: name, session: newSession()}
}

// Next devolve a próxima ref da sessão, no formato <nome>-<sessão>-<n>.
// n cresce a cada chamada.
func (r *Refs) Next() string {
	r.n++
	return fmt.Sprintf("%s-%s-%d", r.name, r.session, r.n)
}

func newSession() string {
	const alpha = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, 4)
	for i := range b {
		b[i] = alpha[rand.Intn(len(alpha))]
	}
	return string(b)
}

// EncodeHello monta a linha de apresentação do cliente.
//
// role é "colony" (leva o name) ou "spectator" (sem name).
//
// Exemplo:
//
//	EncodeHello("aurora-k3f9-1", "colony", "aurora")
//	// "HELLO v1 aurora-k3f9-1 colony aurora"
func EncodeHello(ref, role, name string) string {
	if role == "spectator" {
		return fmt.Sprintf("HELLO v1 %s spectator", ref)
	}
	return fmt.Sprintf("HELLO v1 %s colony %s", ref, name)
}

// E2: ação de domínio -> linha de protocolo com ref correto.
// TODO E2: implemente. O formato é:
//
//	ACT <ref> <kind> <x> <y>   para expand, attack e fortify
//	ACT <ref> pass             para pass (sem coordenadas)
//
// A ação de domínio é Action{Kind: "expand" | "attack" | "fortify", X: x, Y: y}
// ou Action{Kind: "pass"}.
// Enquanto o stub estiver aqui a sua colônia só passa a vez.
func EncodeAction(a Action, ref string) string {
	return fmt.Sprintf("ACT %s pass", ref)
}

// EncodePing monta a linha "PING <ref>", usada para medir latência.
func EncodePing(ref string) string {
	return fmt.Sprintf("PING %s", ref)
}

// --- parse -----------------------------------------------------------------

func atoi(tok string) (int, bool) {
	n, err := strconv.Atoi(tok)
	return n, err == nil
}

func tok(t []string, i int) string {
	if i < len(t) {
		return t[i]
	}
	return ""
}

// ParseCell converte um token de célula em uma Cell de domínio.
//
// O formato do token é "x,y,terreno,dono,fortificada"
// (ex.: "9,5,plain,3,1"). Tokens extras na célula são ignorados.
//
// Devolve a célula e true; se o token for malformado, devolve Cell{} e false.
//
// Exemplo:
//
//	c, ok := ParseCell("9,5,plain,3,1")
//	// c == Cell{X: 9, Y: 5, Terrain: "plain", Owner: 3, Fortified: 1}, ok == true
func ParseCell(tok string) (Cell, bool) {
	f := strings.Split(tok, ",")
	if len(f) < 5 {
		return Cell{}, false
	}
	x, ok1 := atoi(f[0])
	y, ok2 := atoi(f[1])
	owner, ok3 := atoi(f[3])
	fortified, ok4 := atoi(f[4])
	if !ok1 || !ok2 || !ok3 || !ok4 {
		return Cell{}, false
	}
	return Cell{X: x, Y: y, Terrain: f[2], Owner: owner, Fortified: fortified}, true
}

func parseCellList(tok string) ([]Cell, bool) {
	if tok == "" {
		return nil, true
	}
	parts := strings.Split(tok, ";")
	cells := make([]Cell, 0, len(parts))
	for _, p := range parts {
		c, ok := ParseCell(p)
		if !ok {
			return nil, false
		}
		cells = append(cells, c)
	}
	return cells, true
}

// E1: uma linha OBS crua -> observação de domínio.
// TODO E1: implemente. O formato da linha é:
//
//	OBS <ref> <tick> <status> <scores_tick> <celulas>
//
// status é "alive" ou "dead"; celulas é a lista separada por ";"
// (ParseCell e parseCellList acima já existem, use-as). Devolva:
//
//	Message{Type: "obs", Ref: ..., Tick: ..., Status: ..., ScoresTick: ..., Cells: ...}
//
// Linha malformada devolve erro, nunca panica. Tokens extras no final
// são ignorados (é o que salva o seu cliente no drill da v2).
func ParseObservation(line string) (Message, error) {
	return Message{}, fmt.Errorf("TODO E1: implemente ParseObservation")
}

func parseWelcome(t []string) (Message, error) {
	if tok(t, 2) == "spectator" {
		w, ok1 := atoi(tok(t, 3))
		h, ok2 := atoi(tok(t, 4))
		tickMs, ok3 := atoi(tok(t, 5))
		if !ok1 || !ok2 || !ok3 {
			return Message{}, fmt.Errorf("welcome de espectador malformado")
		}
		cells, ok := parseCellList(tok(t, 6))
		if !ok {
			return Message{}, fmt.Errorf("snapshot malformado")
		}
		return Message{
			Type: "welcome", Role: "spectator", W: w, H: h, TickMs: tickMs, Cells: cells,
		}, nil
	}
	id, ok1 := atoi(tok(t, 2))
	w, ok2 := atoi(tok(t, 5))
	h, ok3 := atoi(tok(t, 6))
	tickMs, ok4 := atoi(tok(t, 7))
	viewRadius, ok5 := atoi(tok(t, 8))
	if !ok1 || !ok2 || !ok3 || !ok4 || !ok5 {
		return Message{}, fmt.Errorf("welcome malformado")
	}
	spawn := strings.Split(tok(t, 9), ",")
	if len(spawn) < 2 {
		return Message{}, fmt.Errorf("spawn malformado")
	}
	sx, ok6 := atoi(spawn[0])
	sy, ok7 := atoi(spawn[1])
	if !ok6 || !ok7 {
		return Message{}, fmt.Errorf("spawn malformado")
	}
	return Message{
		Type: "welcome", Role: "colony", ID: id, Name: tok(t, 3), Color: tok(t, 4),
		W: w, H: h, TickMs: tickMs, ViewRadius: viewRadius, SpawnX: sx, SpawnY: sy,
	}, nil
}

func parseScore(t []string) (Message, error) {
	tick, ok := atoi(tok(t, 2))
	if !ok {
		return Message{}, fmt.Errorf("score sem tick")
	}
	var entries []ScoreEntry
	for _, e := range strings.Split(tok(t, 3), ";") {
		if e == "" {
			continue
		}
		f := strings.Split(e, ",")
		id, ok1 := atoi(tok(f, 0))
		cells, ok2 := atoi(tok(f, 2))
		if !ok1 || !ok2 {
			return Message{}, fmt.Errorf("score malformado")
		}
		entries = append(entries, ScoreEntry{
			ID: id, Name: tok(f, 1), Cells: cells, Status: tok(f, 3),
		})
	}
	return Message{Type: "score", Ref: tok(t, 1), Tick: tick, Entries: entries}, nil
}

func parseDiff(t []string) (Message, error) {
	tick, ok := atoi(tok(t, 2))
	if !ok {
		return Message{}, fmt.Errorf("diff sem tick")
	}
	var changes []Change
	for _, c := range strings.Split(tok(t, 3), ";") {
		if c == "" {
			continue
		}
		f := strings.Split(c, ",")
		x, ok1 := atoi(tok(f, 0))
		y, ok2 := atoi(tok(f, 1))
		owner, ok3 := atoi(tok(f, 2))
		fortified, ok4 := atoi(tok(f, 3))
		if !ok1 || !ok2 || !ok3 || !ok4 {
			return Message{}, fmt.Errorf("diff malformado")
		}
		changes = append(changes, Change{X: x, Y: y, Owner: owner, Fortified: fortified})
	}
	return Message{Type: "diff", Ref: tok(t, 1), Tick: tick, Changes: changes}, nil
}

// ParseLine despacha uma linha crua do servidor para o parse do tipo certo.
//
// Cobre WELCOME, OBS, ACK, NACK, ERR, SCORE, DIFF e PONG. O campo Type da
// Message devolvida discrimina o tipo. Linha malformada devolve erro.
//
// Exemplo:
//
//	msg, err := ParseLine("ACK aurora-k3f9-1 97")
//	// err == nil, msg.Type == "ack", msg.Ref == "aurora-k3f9-1", msg.Tick == 97
func ParseLine(line string) (Message, error) {
	t := strings.Split(line, " ")
	switch t[0] {
	case "OBS":
		return ParseObservation(line)
	case "WELCOME":
		return parseWelcome(t)
	case "ACK":
		tick, ok := atoi(tok(t, 2))
		if !ok {
			return Message{}, fmt.Errorf("ack sem tick")
		}
		return Message{Type: "ack", Ref: tok(t, 1), Tick: tick}, nil
	case "NACK":
		if len(t) < 3 {
			return Message{}, fmt.Errorf("nack malformado")
		}
		return Message{
			Type: "nack", Ref: t[1], Code: t[2], Detail: strings.Join(t[3:], " "),
		}, nil
	case "ERR":
		if len(t) < 3 {
			return Message{}, fmt.Errorf("err malformado")
		}
		return Message{
			Type: "err", Ref: t[1], Code: t[2], Detail: strings.Join(t[3:], " "),
		}, nil
	case "SCORE":
		return parseScore(t)
	case "DIFF":
		return parseDiff(t)
	case "PONG":
		return Message{Type: "pong", Ref: tok(t, 1)}, nil
	default:
		return Message{}, fmt.Errorf("tipo desconhecido %s", t[0])
	}
}

File: client-c/src/protocol.c
#include "protocol.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

/* buffer de tokenização: parse_* destroem uma cópia da linha.
 * ponytail: static, o cliente é single-thread */
#define SCRATCH (256 * 1024)

static const char ALPHA[] = "abcdefghijklmnopqrstuvwxyz0123456789";

/* --- encode --------------------------------------------------------------- */

void refs_init_session(Refs *refs, const char *name, const char *session) {
  snprintf(refs->name, sizeof refs->name, "%s", name);
  snprintf(refs->session, sizeof refs->session, "%s", session);
  refs->n = 0;
}

void refs_init(Refs *refs, const char *name) {
  char session[5];
  int i;
  srand((unsigned)(time(NULL) ^ ((long)getpid() << 16)));
  for (i = 0; i < 4; i++)
    session[i] = ALPHA[rand() % (sizeof ALPHA - 1)];
  session[4] = '\0';
  refs_init_session(refs, name, session);
}

void refs_next(Refs *refs, char out[MAX_REF]) {
  snprintf(out, MAX_REF, "%s-%s-%ld", refs->name, refs->session, ++refs->n);
}

void encode_hello(char *out, size_t cap, const char *ref, const char *name) {
  snprintf(out, cap, "HELLO v1 %s colony %s", ref, name);
}

void encode_action(char *out, size_t cap, const Action *action, const char *ref) {
  if (strcmp(action->kind, "pass") == 0) {
    snprintf(out, cap, "ACT %s pass", ref);
  } else {
    snprintf(out, cap, "ACT %s %s %d %d", ref, action->kind, action->x, action->y);
  }
}

void encode_ping(char *out, size_t cap, const char *ref) {
  snprintf(out, cap, "PING %s", ref);
}

/* --- parse ---------------------------------------------------------------- */

static int err(Msg *out, const char *reason) {
  out->type = MSG_ERROR;
  snprintf(out->error, sizeof out->error, "%s", reason);
  return -1;
}

/* inteiros base-10; qualquer outra coisa torna a linha malformada.
 * Rejeita "+5" e lixo depois do número, como os outros clientes. */
static int to_int(const char *s, long *out) {
  char *end;
  long v;
  if (!s || !*s || *s == '+')
    return -1;
  v = strtol(s, &end, 10);
  if (end == s || *end != '\0')
    return -1;
  *out = v;
  return 0;
}

/* divide `s` in-place por `delim`; devolve quantos tokens couberam em `toks` */
static int split(char *s, char delim, char *toks[], int max) {
  int n = 0;
  toks[n++] = s;
  for (; *s && n < max; s++) {
    if (*s == delim) {
      *s = '\0';
      toks[n++] = s + 1;
    }
  }
  return n;
}

/* célula: x,y,terrain,owner,fortified (campos extras na célula ignorados) */
int parse_cell(const char *tok, Cell *out) {
  char buf[64];
  char *f[8];
  long x, y, owner, fortified;

  if (strlen(tok) >= sizeof buf)
    return -1;
  strcpy(buf, tok);
  if (split(buf, ',', f, 8) < 5)
    return -1;
  if (to_int(f[0], &x) || to_int(f[1], &y) || to_int(f[3], &owner) ||
      to_int(f[4], &fortified))
    return -1;

  out->x = (int)x;
  out->y = (int)y;
  snprintf(out->terrain, sizeof out->terrain, "%s", f[2]);
  out->owner = (int)owner;
  out->fortified = (int)fortified;
  return 0;
}

/* tok já tokenizado in-place; NULL ou "" = lista vazia */
static int parse_cells(char *tok, Msg *out) {
  char *items[MAX_CELLS];
  int n, i;

  if (!tok || !*tok) {
    out->n_cells = 0;
    return 0;
  }
  n = split(tok, ';', items, MAX_CELLS);
  for (i = 0; i < n; i++) {
    if (parse_cell(items[i], &out->cells[i]))
      return -1;
  }
  out->n_cells = n;
  return 0;
}

/* E1: uma linha OBS crua -> observação de domínio. */
int parse_observation(const char *line, Msg *out) {
  static char buf[SCRATCH];
  char *t[8];
  long tick, scores_tick;
  int n;

  if (strlen(line) >= sizeof buf)
    return err(out, "linha longa demais");
  strcpy(buf, line);

  n = split(buf, ' ', t, 8);
  if (n < 5 || strcmp(t[0], "OBS") != 0)
    return err(out, "nao e uma linha OBS");
  if (to_int(t[2], &tick) || to_int(t[4], &scores_tick))
    return err(out, "tick ausente ou invalido");
  if (strcmp(t[3], "alive") != 0 && strcmp(t[3], "dead") != 0)
    return err(out, "status invalido");

  out->type = MSG_OBS;
  snprintf(out->ref, sizeof out->ref, "%s", t[1]);
  out->tick = tick;
  snprintf(out->status, sizeof out->status, "%s", t[3]);
  out->scores_tick = scores_tick;
  /* n >= 6 ? t[5] : lista vazia; tokens além de t[5] ignorados (leitor
   * tolerante: é o que salva o cliente no drill da v2) */
  if (parse_cells(n >= 6 ? t[5] : NULL, out))
    return err(out, "lista de celulas malformada");
  return 0;
}

/* WELCOME srv-0 3 aurora 96CDFB 60 40 1000 3 4,34 (só colônia) */
static int parse_welcome(char *buf, Msg *out) {
  char *t[12];
  char *spawn[4];
  long id, w, h, tick_ms, view_radius, sx, sy;

  if (split(buf, ' ', t, 12) < 10)
    return err(out, "welcome malformado");
  if (strcmp(t[2], "spectator") == 0)
    return err(out, "este cliente so joga como colonia");
  if (to_int(t[2], &id) || to_int(t[5], &w) || to_int(t[6], &h) ||
      to_int(t[7], &tick_ms) || to_int(t[8], &view_radius))
    return err(out, "welcome malformado");
  if (split(t[9], ',', spawn, 4) < 2 || to_int(spawn[0], &sx) || to_int(spawn[1], &sy))
    return err(out, "spawn malformado");

  out->type = MSG_WELCOME;
  out->id = (int)id;
  snprintf(out->name, sizeof out->name, "%s", t[3]);
  snprintf(out->color, sizeof out->color, "%s", t[4]);
  out->w = (int)w;
  out->h = (int)h;
  out->tick_ms = (int)tick_ms;
  out->view_radius = (int)view_radius;
  out->spawn_x = (int)sx;
  out->spawn_y = (int)sy;
  return 0;
}

static int parse_ack(char *buf, Msg *out) {
  char *t[4];
  long tick;

  if (split(buf, ' ', t, 4) < 3 || to_int(t[2], &tick))
    return err(out, "ack sem tick");

  out->type = MSG_ACK;
  snprintf(out->ref, sizeof out->ref, "%s", t[1]);
  out->tick = tick;
  return 0;
}

/* NACK/ERR <ref> <code> <texto livre até o fim da linha> */
static int parse_nack(char *buf, Msg *out, MsgType type) {
  char *t[16];
  int n, i;
  size_t len = 0;

  n = split(buf, ' ', t, 16);
  if (n < 3)
    return err(out, "nack/err malformado");

  out->type = type;
  snprintf(out->ref, sizeof out->ref, "%s", t[1]);
  snprintf(out->code, sizeof out->code, "%s", t[2]);
  out->detail[0] = '\0';
  for (i = 3; i < n && len + 1 < sizeof out->detail; i++) {
    int wrote = snprintf(out->detail + len, sizeof out->detail - len, "%s%s",
                         i > 3 ? " " : "", t[i]);
    if (wrote < 0)
      break;
    len += (size_t)wrote < sizeof out->detail - len ? (size_t)wrote
                                                    : sizeof out->detail - len - 1;
  }
  return 0;
}

static int parse_score(char *buf, Msg *out) {
  char *t[4];
  char *entries[MAX_SCORE_ENTRIES];
  long tick;
  int n, i;

  if (split(buf, ' ', t, 4) < 3 || to_int(t[2], &tick))
    return err(out, "score sem tick");

  out->type = MSG_SCORE;
  snprintf(out->ref, sizeof out->ref, "%s", t[1]);
  out->tick = tick;
  out->n_entries = 0;

  n = t[3] ? split(t[3], ';', entries, MAX_SCORE_ENTRIES) : 0;
  for (i = 0; i < n; i++) {
    char *f[5];
    long id, cells;
    if (!*entries[i])
      continue;
    if (split(entries[i], ',', f, 5) < 4 || to_int(f[0], &id) || to_int(f[2], &cells))
      return err(out, "score malformado");
    out->entries[out->n_entries].id = (int)id;
    snprintf(out->entries[out->n_entries].name, MAX_NAME, "%s", f[1]);
    out->entries[out->n_entries].cells = (int)cells;
    snprintf(out->entries[out->n_entries].status, 8, "%s", f[3]);
    out->n_entries++;
  }
  return 0;
}

int parse_line(const char *line, Msg *out) {
  static char buf[SCRATCH];

  if (strlen(line) >= sizeof buf)
    return err(out, "linha longa demais");
  strcpy(buf, line);

  if (strncmp(buf, "OBS ", 4) == 0)
    return parse_observation(line, out);
  if (strncmp(buf, "WELCOME ", 8) == 0)
    return parse_welcome(buf, out);
  if (strncmp(buf, "ACK ", 4) == 0)
    return parse_ack(buf, out);
  if (strncmp(buf, "NACK ", 5) == 0)
    return parse_nack(buf, out, MSG_NACK);
  if (strncmp(buf, "ERR ", 4) == 0)
    return parse_nack(buf, out, MSG_ERR);
  if (strncmp(buf, "SCORE ", 6) == 0)
    return parse_score(buf, out);
  if (strncmp(buf, "PONG", 4) == 0) {
    char *t[3];
    out->type = MSG_PONG;
    if (split(buf, ' ', t, 3) >= 2)
      snprintf(out->ref, sizeof out->ref, "%s", t[1]);
    return 0;
  }
  return err(out, "tipo desconhecido");
}
